### ISSUE TITLE
stickyToolbar & overflowed scrollableContainer not working as intended

### ISSUE THREAD

**Issue by Mboulianne** [froala/wysiwyg-editor#2955]
##### Expected behavior.
The toolbar should always stick if a scrollableContainer is specified.

##### Actual behavior.
When there the page and the scrollableContainer have a scrollbar, the toolbar won't stick as it should if the page scrollbar is at its top-most or bottom-most position.

##### Steps to reproduce the problem.

1- Make sure page scrollbar is at its topmost position.
2- Scroll the scrollbar in the scrollcontainer
3- Toolbar didn't stick :(
4- Do the same with scrollbar at its bottom-most position.
5- Toolbar didn't stick.
6- Put the page scrollbar somwhere between its top-most position and its bottom-most position.
7- Scroll the scrollbar in the scroll container.
8 - Toolbar did stick!!!

See jsfiddle https://jsfiddle.net/ppyLhnqt/65/

To make sure everything is clear I'll put some screenshot
1- Page scrollbar is at its top-most position.
![image](https://user-images.githubusercontent.com/18729439/42590197-217dc0fc-8511-11e8-8396-c348fb0f3aaa.png)
The toolbar isn't visible anymore.  It didn't stick.

2- Page scrollbar is at its bottom-most position
![image](https://user-images.githubusercontent.com/18729439/42590229-39169aa4-8511-11e8-940a-88e59c51d314.png)
The toolbar isn't visible anymore.  It didn't stick.

3- Page scrollbar is somwhere in the middle
![image](https://user-images.githubusercontent.com/18729439/42590249-475b42e0-8511-11e8-809a-4650a7106cf0.png)
The toolbar is visible.  It did stick.

##### OS.
Windows 10 

##### Browser.
Firefox 61.0.1 (64-bit)
Chrome 67.0.3396.99 (Official Build) (64-bit)
Edge Microsoft Edge 42.17134.1.0 / Microsoft EdgeHTML 17.17134

**Comment by Mboulianne** [comment]
@gabra20

**Comment by stefanneculai** [comment]
That is because the height of the view is smaller than 500px and the editor is designed to keep the toolbar at the top when it exceeds the scrollable container. If you would set it to a value smaller than the height of the view all is good: https://jsfiddle.net/froala/ppyLhnqt/70/.

**Comment by Mboulianne** [comment]
new jsfiddle: https://jsfiddle.net/ppyLhnqt/103/

how do you explain that I'm getting that now?
![image](https://user-images.githubusercontent.com/18729439/42634467-4bc30426-85b1-11e8-8d00-1c7f84c0c55b.png)

The toolbar appears below my scrollContainer.  I've hadded content before and after my container to represent my real use case

Also
![image](https://user-images.githubusercontent.com/18729439/42634549-7d96b39e-85b1-11e8-92be-8676ebb9319a.png)
If I only scroll the page (not the scroll container) the toolbar don't stick at all.

It looks pretty buggy to me.

**Comment by stefanneculai** [comment]
Thanks for the further details. Indeed, it appears to be a problem. I updated it accordingly.

**Comment by Mboulianne** [comment]
I think I found a similar bug: 
My editors are contained in a fixed container that lies at the end / bottom of the body tag with top: 0; bottom: 0; left: 0; right: 0; (it take the whole viewport).  If the body has enough content to make a scrollbar appear and if that scrollbar is scrolled, editor toolbar won't stick correctly

jsFiddle:
https://jsfiddle.net/184f7zpd/63/

Body has no scrollbar => 
![image](https://user-images.githubusercontent.com/18729439/42763244-788a61fc-88e0-11e8-9787-ba9fcbabff97.png)
(everything is ok)

Body has a scrollbar but the user haven't scrolled yet =>
![image](https://user-images.githubusercontent.com/18729439/42763268-8e8f331a-88e0-11e8-898a-0315da467a5c.png)
(everything is ok)

Body has a scrollbar and the user have scrolled =>
![image](https://user-images.githubusercontent.com/18729439/42763309-a61b2688-88e0-11e8-8b9f-7f2f44bb8495.png)
Buggy!

**Comment by Mboulianne** [comment]
As a workaround you can do 
if (this.scrollActive) {
            this.scrollActive = false;
            this.stopScrollPosition = this.scrollPosition;
            this.bodyStyle.position = 'fixed';
            this.bodyStyle.top = '0'; // --> ENA2-767
            this.bodyStyle.right = '0';
            this.bodyStyle.left = '0';
            this.bodyStyle.bottom = '0'; // --- Added bug in IE11 --- Added to fix edge case where showed contents through popper/portal are hidden when page content isn't high enough to stretch the body.
            this.bodyStyle.height = '100%';
            this.bodyStyle.overflow = 'hidden';
            this.bodyEl.scrollTop = this.stopScrollPosition;
            this.htmlEl.style.overflow = 'hidden';
        }

when opening the window / modal.  However it breaks the tooltip (they can't position correctly).  It leads me to think that positioning tooltips in a fixed body isn't working either,.

**Comment by stefanneculai** [comment]
Related to https://github.com/froala/wysiwyg-editor/issues/2981.

**Comment by Mboulianne** [comment]
@stefanneculai  are you sure these are related?  My bug is about the toolbar that isn't positionned correctly.  However, I opened a similar bug https://github.com/froala/wysiwyg-editor/issues/3038 ?

**Comment by stefanneculai** [comment]
Yes, they are related because they use the same logic for positioning inside a `scrollableContainer`.

**Comment by Mboulianne** [comment]
@stefanneculai ok we'll see once this get resolved then.  It's not a big deal for us since we found a workaround.

**Comment by loocaworld** [comment]
Is this bug fixed?

**Comment by JessieCowie23** [comment]
Is a fix for this still in the works? I'm having the same issue in 3.0.6

**Comment by Spankyed** [comment]
Such a simple issue faced by many, yet 3 years later still no accepted solution.

**Comment by brucetp** [comment]
This is really a frustrating bug, so I decided to work around it myself.   I'm using Vue but the concept still works in plain JS.  "ActivityPage" is my scrollable DIV.  

1)  Attach to the scroll event listener on the DIV Tag when the page loads
2)  Wait for scrolling, then disable the listener (better performance than constantly checking)
3)  Check to see how far we've scrolled down the page  (my editor is approx 30% down the page, yours will very)
4)  If scrolling is less than 30%, remove the 'sticky' class from the froala toolbar, otherwise add the class because we have scrolled past it
5)  remove the listener when navigating away from the page

[The code block wouldn't format this correctly so that why it's in plan text]

  mounted () {
    document.getElementById('ActivityPage').addEventListener('scroll', this.handleScroll, { passive: true })
  },
  destroyed () {
    document.getElementById('ActivityPage').removeEventListener('scroll', this.handleScroll)
  },
  methods: {
    handleScroll (event) {
      document.getElementById('ActivityPage').removeEventListener('scroll', this.handleScroll)
      console.log('stop listening for performance reasons')
      setTimeout(() => {
        document.getElementById('ActivityPage').addEventListener('scroll', this.handleScroll, { passive: true })
        console.log('start listening')
        this.handleStickyToolbar()
      }, 300)
    },
    handleStickyToolbar () {
      var element = document.getElementById('ActivityPage')
      var b = element.scrollHeight - window.innerHeight
      var c = element.scrollTop / b
      var x = document.getElementsByClassName('fr-toolbar')
      if (c < 0.3) {
        x[0].classList.remove('fr-sticky-on')
      } else {
        x[0].classList.add('fr-sticky-on')
      }
    },

**Comment by loocaworld** [comment]
on every updates, I do this modification.

1. find
t.css("top", S.helpers.getPX(t.data("top")) + i) # S could be L or something. That "i" is problem.
2. modify this to
t.css("top", S.helpers.getPX(t.data("top")) + (S.$sc.css("position") === "fixed" ? S.$sc.position().top : i))

**Comment by nateleavitt** [comment]
Also, having a similar problem. 

@loocaworld - can you reference where in the code you make that change?

**Comment by loocaworld** [comment]
@nateleavitt 
Variables are always changed on every big updates. 
So if you are using froala editor version 4.0.10, 
1. open file "js/froala_editor.min.js"
2. find this 
`L.helpers.getPX(t.data("top")) + i`
3. change like this
`L.helpers.getPX(t.data("top")) + (L.$sc.css("position") === "fixed" ? L.$sc.position().top : i)`

**Comment by OdaNeo** [comment]
I have same problem, I add this css in global and fix this problem
version:  "react-froala-wysiwyg": "3.0.6",

.fr-toolbar.fr-top {
    position: sticky !important;
    z-index: 2 !important;
}

**Comment by loocaworld** [comment]
This problem is caused because the editor is in a scroll-downed fixed container.
If the background parent has a little offset from the top, 
inner elements in the fixed container's offset().top contains parent's offset too.
So, if $sc(=scrollable container) is in a fixed container or is a fixed container, you have to consider that.

in "js/froala_editor.min.js", find below code
`t.css("top", L.helpers.getPX(t.data("top")) + i)`
change that code line to
`t.css("top", L.helpers.getPX(t.data("top")) + (L.$sc.parents().filter(function(){return $(this).css('position') == 'fixed';}).length ? L.$sc.position().top : i))`

**Comment by dplarina** [comment]
I have a similar problem. The scrollable container is not fixed but it sounds like the same issue.

If the scrollable container is offset from the top of the viewport at all, it will offset the toolbar down by that amount.
I created a minimal repro here with the latest version (4.0.19) https://jsfiddle.net/2spd3tha/2/

The body adds 8px. The wrapper element adds 16px, and the header adds 64px, so the scrollable container is pushed down for a total of 8px + 16px + 64px = 88px. Since the offsetTop of the scrollable container is 88px, it sets the top of the toolbar to 88px, despite me setting the toolbarStickyOffset to 0.

![image](https://github.com/froala/wysiwyg-editor/assets/1740636/2d3c66ef-37ad-4a9f-a973-7586f60992e6)

**Comment by shan-du** [comment]
2025, i think this issue still exist, it's quite annoying

**Comment by fexxdev** [comment]
Yeah we use froala in production and we have the same exact issue. Im trying to work-around it, but i would be nice to get a fix in the future.

**Comment by ilyaskarim** [comment]
This issue is being addressed in our V4.5, Please refer: https://froala.com/wysiwyg-editor/changelog/#4.5. 
If you are experiencing this issue, please proceed to upgrade to our latest version, as this will resolve the problem on your end.
